### PR TITLE
Proposal: Send scrapped Metadata as part of remote-write

### PR DESCRIPTION
**Pull request by gotjosh**
Hi all 👋,

First of all, apologies for the long text. I've been chatting with a few people individually (Tom, Callum and Goutham), and I wanted to collate all the feedback I've received into a single place so that we can decide how to best proceed with this change. Callum has been kindly helping me figure this out and he can take as much credit as I do for everything that I'm proposing. 

I'm not going to cover the reasons for doing this as that lives within the [design doc](https://docs.google.com/document/d/1GoWgKuioZhjFcDNyyrs_WrKxhHR8UZNFCTqkpt04k4M/edit?ts=5e237203#). Instead, I'm going to dive down straight into the current state and technical tradeoffs.

## Current state
This is the simplest thing I could build that takes metadata from the scrape cache and sends it through the remote-write protocol. 

It's missing tests in almost all the areas and names of things might be slightly off - but I don't want that to be the focus on this. Instead, I'm looking for feedback on the overarching approach and how this will affect the operational complexity of the protocol.

With that said, let's look at some trade-offs that I would love to decide on.

## 1. Within a `WriteRequest`, a different request (e.g. `MetadataWriteRequest`), or a mix.
 
Within this PR, the metadata information is being sent as part of a `WriteRequest` but on a different field, **all at once, and every time**. The increase of bytes is about ~%10 with a single target (from ~600 bytes to ~660 bytes after compression) but this not have linear growth as metadata across targets overlaps, e.g. 10 targets exposing metrics from `node_exporter` will all have the same metadata.

Screenshot 2020-01-17 at 18 28 03

That being said, **I'm not convinced this should be the way to move forward**. Sending all the metadata at once and every time is an oversimplification done as part of this exercise.

Realistically, we'd need to:

1. determine a fixed ratio of metadata/samples to send on each request, e.g. 10% metadata and 90% series 
2. keep track of which metadata we've sent as part of X batch
3. just before sending figure out if we're not at full capacity for series and use the remainder of the batch to send more metadata. 

On top of that, we'd need to introduce somehow an ewma for keeping track the rate at which we're receiving/dropping metadata and use those plus the series ewma to know when to scale up/down shards. I think this makes the operation of the current approach a bit more complicated than what we might want at this stage.

In the past, I've considered the option of sending it as part of a separate `MetadataWriteRequest`, but that was quickly dismissed due two a) having to hide it behind a flag and have many potential users miss out of the feature and b) integrations needing to implement this functionality as part of a separate endpoint. However, on the development and operational side, this would be more straightforward:

1. We could create new ewma that does not take into account the series/samples. Eliminating the need to come up with a unified heuristic.
2. We could probably get away with sending all the metadata we currently hold all at once (after we've done some measurements) and wouldn't have a need to figure out the ratio between series/metadata.

Speaking with Tom today, I realise I misunderstood his original comment. Being a separate request is not equal to sending metadata and series together. That means we can happily have a separate set of shards (and code path) exclusively for sending metadata and use the same `WriteRequest` with no series data, just metadata in it. 

This would be a little different from what it is today as the protocol does not send _any_ data if we don't have pending samples but seems like a fine tradeoff? As we know, integrations that don't support the new field within `WriteRequest` will receive bytes within the `XXX` fields of the protobuf.

**What do others think?** 

I know Goutham mentioned some sort of content-negotiation ala remote-read, where (if my understanding is correct) we'd negotiate every X amount of time on whenever the client is accepting metadata request - if it is, then we'd start the engine. This would help us CPU/memory by not keeping that part of the protocol running until it's needed.

## 2. How often and how much metadata do we send? 

I understand the answer to this depends on the question above, but I would love to get feedback on what I have in mind.

1. Have a hard limit on the number of metrics we send metadata for at a time, e.g. 5000 (the current number is made up, I will open a PR to Prometheus so that we can measure individual metadata caches sizes both in terms of bytes and number of entries to come up with a better heuristic for a number)
2. Have a deadline of X * 3 where X is the scrape interval, e.g. When the internal is 20s we'd only send metadata every 60s. 
3. Ideally, send them all at once. In the case where we need to split up metadata across multiple requests, I'm open to ideas on what to do there? (Do we send X requests at the same time, do we keep track of the metrics we've sent and then send the remainder later on, etc.)

## 3. Should we include the target as part of the payload?

The end goal of this work is to replicate both `/api/v1/metadata` and `api/v1/target/metadata` from Prometheus within Cortex. For the latter, we need the target information.

There are only two known users of `api/v1/target/metadata`, Stackdriver for their Prometheus sidecar and the PromQL LSP. Stackdriver is never going to use cortex for this as this is specifically for Prometheus, the LSP is currently moving from using this endpoint to the newer one.

Should we not include the target information and thus not implement `api/v1/target/metadata` in Cortex?

Goutham suggested this and at first, I felt a bit uneasy for the sake of completeness. Thinking a bit more about this I do see value in removing it as it would make the operation of pending metadata much simpler than what it is right now (where we need to keep a set of the target's we've seen the metadata at).

### CODE REVIEW

File: scrape/target.go [hunk 1]
```diff
@@ -75,11 +75,18 @@ func (t *Target) String() string {
...
 }
 
+// MetricMetadataStore represents a storage for metadata.
 type MetricMetadataStore interface {
 	ListMetadata() []MetricMetadata
 	GetMetadata(metric string) (MetricMetadata, bool)
 }
 
+// MetricMetadataAppender provides metadata appends by target against a storage.
+type MetricMetadataAppender interface {
```
Comment (gotjosh) [inline scrape/target.go:85]: https://github.com/gotjosh/prometheus/blob/2b2eb79e8b7245d2d60001ae08443d31f7b3a5b9/storage/interface.go#L90-L99

This interface probably belongs in the same place as the storage one, but the small problem here is that both `Target` and `MetricMetadata` belong to the `scrape` package which is not imported as part of storage - and it feels it shouldn't be. Targets we could probably get by just using `labels.Labels` of the target but I don't see a way for `MetricMetadata`, am I missing something?

The trade-off I see is the allocation of new structs (which would belong to the storage package) vs having this interface in a bit of an odd place - I took the latter. Another approach would be an interface for `MetricMetadata` but that seems too complex for this (and not a pattern I've seen used in the codebase)

File: scrape/scrape.go [hunk 2]
```diff
@@ -946,6 +963,10 @@ mainLoop:
 		}
 		last = start
 
+		// Now that we're done scraping. Let's append metadata to remote-write.
+		// Append metadata to remote-write
+		ma.Set(sl.target, sl.cache.ListMetadata())
```
Comment (gotjosh) [inline scrape/scrape.go:968]: Metadata is set incrementally in the cache as it is parsed: first HELP, then Type, etc. Sadly, this makes it impossible to set the metadata as it scraped to remote-write. We need to wait until we're done scraping the target before sending it off the entire metadata cache for this target.

**Comment by beorn7** [comment]
I'll leave the code-level review to experts of the remote-write part.

About the questions raised above:

> Should we include the target as part of the payload?

I would indeed not, at least initially. As you already concluded, the only user of the target-specific API is Stackdriver, and I doubt they'll ever use Cortex.

Let's only go for reproducing the `/api/v1/metadata` for now, including it's handling of ambiguity (multiple sets of metadata for the same metric name).

> How often and how much metadata do we send?

This is really hard. I assume we have to act within the constraints of a fully push-based data delivery, i.e. there is not way the remote-write receiver can communicate back any needs.

Ideally, a newly arriving remote-write receiver would get all the metadata initially, and then it would get only changes. This would be more involved on the Prometheus side (the new engine needs to get notified if a target gets added or its metadata changes), and I guess we cannot even rely on a receiver never needing a complete update ever again.

Limiting metadata to a certain low percentage of the total throughput sounds like a good idea, but on the other hand we have to make sure to send it even in low-traffic situation (have a minimal absolute throughput for metadata, too?).

I like the idea of doing metadata best-effort and not send it if we are behind on pending sample data.

File: scrape/target.go (hunk 1 above)
Comment (cstyan) [reply scrape/target.go:85]: I think it makes sense to have this interface live alongside the existing `Appender` interface. There's already usage of `scrape.Target` by the web code, you could do the same, or consider moving target related data into a new package? Perhaps we should have a metadata package?

**Comment by cstyan** [comment]
My thoughts, very loosely organized.

For 1:

> Speaking with Tom today, I realise I misunderstood his original comment. Being a separate request is not equal to sending metadata and series together. That means we can happily have a separate set of shards (and code path) exclusively for sending metadata and use the same WriteRequest with no series data, just metadata in it.

That's one way to get around having to come up with a way to calculate # of shards from both samples and metadata rates, which was my main reason for leaning towards separate requests.

For 2:

I'm not really understanding any of the 3 options. The metadata proto type contains all the info needed for the metadata right, there's no reason we have to send metadata only for the series that are within the request? Assuming we continue with having metadata and samples in the same request (not just the same request type).

If we split metadata out into separate requests, we should just come up with a way to have metadata be sent as it's updated within some reasonable amount of time. So obviously at the start we're going to have a bunch of new metadata for each scrape. After that, we only have the list of targets for a given metadata updating infrequently, or series added/removed even more infrequently. Is there something I'm missing?

For 3:

I think we should even extend this and implement a metadata endpoint that takes a PromQL query, similar to what we're planning with exemplars. As you and I talked about earlier, Grafana doesn't even pass a label matcher when it makes a metadata query because the authors don't want to write their own PromQL parser.


Anyways, @beorn7 's pointed out the pros and cons of different strategies for sending, whether only sending updates or always sending everything. I think `I guess we cannot even rely on a receiver never needing a complete update ever again.` is a good argument for always sending all metadata. 
The sharding calculation could be something like:
- snapshot current metadata and determine size
- based on bytes/s throughput of current samples shards implementation, determine how many we would need in order to send all metadata over in X # of s, where X is configurable by the user but defaults to something sane like 1m

**Comment by beorn7** [comment]
> I think we should even extend this and implement a metadata endpoint that takes a PromQL query, similar to what we're planning with exemplars. As you and I talked about earlier, Grafana doesn't even pass a label matcher when it makes a metadata query because the authors don't want to write their own PromQL parser.

I don't think we can compare exemplars and metadata at this point.

In Prometheus, metadata is, as of now, bound to the combination of metric name (or the metric family, if you like that nomenclature) and the target. Grafana and LSP aren't interested in targets, so we offer that flattened API where metadata (one or more sets of which) is bound just to a metric name.

Exemplars are bound to specific buckets of a specific histogram (a metric child) or to a specific counter (again a metric child). That's why they have to be retrieved with whole label matchers rather than just a metric name.

Thus, the nature of the `/api/v1/metadata` endpoint is due to the current data model, not due to any laziness of the Grafana devs.

Personally, I believe we should eventually have metadata per metric child and also persisted over time (let's call it "per series metadata"). But that's just a vision at the moment. For now, the `/api/v1/metadata` is the state of the art, and that's what Cortex needs to be serving.

File: scrape/target.go (hunk 1 above)
Comment (gotjosh) [reply scrape/target.go:85]: > Perhaps we should have a metadata package?

I think this is too big of a change to be included as part of this. I'll keep in mind to suggest it as a follow-up.

The problem here is that the scrape code imports the storage package, attempting to import anything from the scrape package into storage results in a `import cycle not allowed`

**Comment by gotjosh** [comment]
Within Prometheus, Pull Request #6675 adds metrics to the metadata cache by target.

**Comment by cstyan** [comment]
> Thus, the nature of the /api/v1/metadata endpoint is due to the current data model, not due to any laziness of the Grafana devs.

I wasn't trying to suggest they are lazy, more pointing out that no one wants to, and it's unreasonable to expect them to, maintain their own PromQL parser. Even if the existing parser is something we said was stable for external use, it's only in Go. That's not an issue for us in this case but still, no one other than Prometheus team members want to maintain a parser.

So, if any API endpoints that may be used alongside an existing query to get additional data can take a PromQL query instead of only a list of label matchers, that's beneficial to end users.

**Pull request closed without merging**